### PR TITLE
Do not listen to changes made through the Options API before init.

### LINKED ISSUE #52435: WordPress 6.7: PHP Notice on _load_textdomain_just_in_time when updating options before init
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

After the latest changes in WordPress 6.7 (RC1 and RC2), when a third party updates an option before `init` in a site in a non-default language, WooCommerce produces a PHP notice related to the usage of `_load_textdomain_just_in_time`.

### Expected behavior

WooCommerce should not produce such PHP notice.

### Actual behavior

A PHP notice is generated, with the debug message `PHP Notice:  Function _load_textdomain_just_in_time was called <strong>incorrectly</strong>. Translation loading for the <code>woocommerce</code> domain was triggered too early.`.

### Steps to reproduce

1. Get a clean WordPress installation using WordPress 6.7 RC1 or RC2. Make sure that the site language is different than the default - French should do it.
2. Install WooCommerce 9.4.0 RC2.
3. Get a thbird party to update an option before `init`. This following snippet in a mu-plugin should do it:

```
add_action( 'plugins_loaded', function() {
	update_option( 'foo', 'bar' );
} );
```
4. Login and load the site backend. A PHP notice will be gnerated.

### WordPress Environment

<details><summary>System Status Report</summary>

```
### WordPress Environment ###

WordPress address (URL): [Redacted]
Site address (URL): [Redacted]
WC Version: 9.4.0-rc.2
Legacy REST API Package Version: L’extension Legacy REST API n’est pas installée sur ce site.
Action Scheduler Version: ✔ 3.8.2
Log Directory Writable: ✔
WP Version: 6.7-RC2
WP Multisite: –
WP Memory Limit: 256 Mo
WP Debug Mode: ✔
WP Cron: –
Language: fr_FR
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.52 (Ubuntu)
PHP Version: 7.4.33
PHP Post Max Size: 0 o
PHP Time Limit: 300
PHP Max Input Vars: 1000
cURL Version: 7.81.0
OpenSSL/3.0.2

SUHOSIN Installed: –
MySQL Version: 8.0.39-0ubuntu0.22.04.1
Max Upload Size: 0 o
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ❌ Votre serveur n’a pas la classe SoapClient activée - certaines extensions de passerelle utilisant SOAP peuvent ne pas fonctionner correctement.
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

[REDACTED]

### Post Type Counts ###

attachment: 2
page: 8
post: 2

### Security ###

Secure connection (HTTPS): ❌
					Votre boutique n’utilise pas HTTPS. En savoir plus sur HTTPS et les certificats SSL.
Hide errors from visitors: ❌Les messages d’erreur ne doivent pas s’afficher aux visiteurs.

### Active Plugins (2) ###

WooCommerce: par Automattic – 9.4.0-rc.2

### Inactive Plugins (179) ###

### Dropin Plugins () ###

### Must Use Plugins (7) ###

### Settings ###

Legacy API Enabled: ✔
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: –
HPOS feature enabled: –
Order datastore: WC_Order_Data_Store_CPT
HPOS data sync enabled: –

### Logging ###

Enabled: ✔
Handler: Automattic\WooCommerce\Internal\Admin\Logging\LogHandlerFileV2
Retention period: 30 jours
Level threshold: –
Log directory size: 729 Ko

### WC Pages ###

Base de la boutique: #25 - /shop/
Panier: #26 - /cart/ -  Contient le code court [woocommerce_cart]
Commander: #27 - /checkout/ -  Contient le code court [woocommerce_checkout]
Mon compte: #28 - /my-account/
Conditions générales de vente et d’utilisation: ❌ Page non définie

### Theme ###

Name: Storefront
Version: 4.6.0
Author URL: https://woocommerce.com/
Child Theme: ❌ – Si vous modifiez WooCommerce sur un thème parent que vous n‘avez pas construit personnellement
nous vous recommandons d‘utiliser un thème enfant. Voir : Comment créer un thème enfant

WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
product-block-editor
coupons
core-profiler
customize-store
customer-effort-score-tracks
import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
pattern-toolkit-full-composability
product-custom-fields
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
printful
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page
launch-your-store

Disabled Features: product-data-views
experimental-blocks
minified-js
product-pre-publish-modal
settings
async-product-editor-category-field
product-editor-template-system
blueprint
reactify-classic-payments-settings
use-wp-horizon

Daily Cron: ✔ Next scheduled: 2021-04-12 12:12:44 +00:00
Options: ✔
Notes: 79
Onboarding: completed

### Action Scheduler ###

Terminé: 8
Oldest: 2024-10-29 21:29:35 +0000
Newest: 2024-10-29 21:54:22 +0000

En attente: 3
Oldest: 2024-10-29 22:29:35 +0000
Newest: 2024-11-05 21:29:37 +0000


### Status report information ###

Generated at: 2024-10-29 22:06:32 +00:00
```
</details>

### Isolating the problem

- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/products/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by barryhughes**
Improves compatibility with WP 6.7 (and changes to l10n/i18n handling in that version), by limiting when we allow the list of feature definitions to be created.

---

Features can be enabled or disabled by creating or changing an option (via the [WordPress Options API](https://developer.wordpress.org/apis/options/)). For example, the following enables our Analytics feature:

```php
update_option( 'woocommerce_analytics_enabled', 'yes' );
```

To support this, WooCommerce uses filter hooks to listen for options being created or updated. To determine if the option is related to a WooCommerce feature, one part of its process is to look at a list of feature definitions. Building this list involves translating various strings (feature names, etc) which, as of WP 6.7, is frowned upon (see discussion in linked issue) if it takes place before `init`.

It seems quite reasonable that features would not be enabled or disabled before `init`, however, so in this change we only add our event listeners *after* `init` has started (in other words, on `init` priority `0`).

Closes #52435.

### How to test the changes in this Pull Request:

Setup: Make sure your site site is running the latest release candidate of [WordPress 6.7](https://make.wordpress.org/core/6-7/). Then start by installing and activating a non-default language. It doesn't especially matter which, but let's use French here:

```sh
wp core language install fr_FR
wp site switch-language fr_FR
wp language plugin update woocommerce
```

We are going to be looking out for *doing it wrong* notices. A handy means of detecting these is to install and activate [Query Monitor](https://wordpress.org/plugins/query-monitor/), but you can use any method you are familiar and comfortable with. Next, add the following code in a suitable location, such as `wp-content/mu-plugins/fix-52435.php`:

```php
<?php

add_action(
	'plugins_loaded',
	fn () => update_option( 'unrelated_option', uniqid() )
);
```

1. Using current `trunk` code (ie, without this change) refresh any frontend page. You should observe a *doing it wrong* error referencing `_load_textdomain_just_in_time` and **also referencing the woocommerce text domain**:

<div align="center">'Doing it wrong' error about translations, captured by Query Monitor</div>

3. Switch to this branch and repeat. The error should go away, but you may see very similar errors. For the purposes of this PR, that's fine so long as they do not reference the **woocommerce** text domain.

Lastly, but importantly, let's verify we can still enable and disable features by a couple of different channels. With this branch checked out:

1. Visit **WooCommerce ‣ Settings ‣ Advanced ‣ Features** and enable **Analytics** (or disable it, if it was already enabled).
2. Repeat in the opposite direction (disable it, or re-enable it, to get back to your original status).
3. The **Analytics** admin menu option should appear and disappear in alignment with the changes you made.
4. Now let's do the same, but using the Options API. We'll take advantage of WP CLI for this. Assuming we start by enabling:

```sh
wp option set woocommerce_analytics_enabled yes
```

5. Refresh the admin area; you should see that the **Analytics** menu item is present. Now disable:

```sh
wp option set woocommerce_analytics_enabled no
```

6. The **Analytics** menu item should disappear. You can of course do these tests in the other direction.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php [hunk 1]
```diff
@@ -89,8 +91,8 @@ class FeaturesController {
 	 * Creates a new instance of the class.
 	 */
 	public function __construct() {
-		self::add_filter( 'updated_option', array( $this, 'process_updated_option' ), 999, 3 );
-		self::add_filter( 'added_option', array( $this, 'process_added_option' ), 999, 3 );
+		self::add_filter( 'init', array( $this, 'start_listening_for_option_changes' ), 10, 0 );
+
 		self::add_filter( 'woocommerce_get_sections_advanced', array( $this, 'add_features_section' ), 10, 1 );
```
Comment (coreymckrill) [inline plugins/woocommerce/src/Internal/Features/FeaturesController.php]:
```suggestion
		self::add_filter( 'init', array( $this, 'start_listening_for_option_changes' ), 10, 0 );
		self::add_filter( 'woocommerce_get_sections_advanced', array( $this, 'add_features_section' ), 10, 1 );
```

**Review by coreymckrill — APPROVED** [review]
👍 Followed all testing steps, and everything worked as expected, via the admin UI and via wp-cli.

Approving, because this approach seems sufficient, but here's one idea to ponder: if for some reason some other extension programmatically updates one of the feature options before `init`, it seems like that would potentially break things in a way that would be difficult to troubleshoot, because the option value would _look_ correct, but the action hook for the corresponding feature would not have fired, so other steps might be left undone. What if, instead of waiting until `init` to start listening, we have a check at the start of `process_updated_option` to see if `init` has fired yet or not, and if not, we run a `doing_it_wrong` and then bail on the rest of the function? The result would basically be the same, I think, but with the addition of some feedback for the developer.

File: plugins/woocommerce/src/Internal/Features/FeaturesController.php (hunk 1 above)
Comment (barryhughes) [reply plugins/woocommerce/src/Internal/Features/FeaturesController.php]: Woops, thanks!

**Comment by barryhughes** [comment]
> What if, instead of waiting until init to start listening, we have a check at the start of process_updated_option to see if init has fired yet or not, and if not, we run a doing_it_wrong and then bail on the rest of the function?

Yep, that was my first thought [back here](https://github.com/woocommerce/woocommerce/issues/52435#issuecomment-2448590129) (see snippet of code we'd add at the top of the method). Reasons I didn't follow that approach:

- In a sense, we would be trading one *doing it wrong* (issued by WP Core/the translation layer) for another (issued by us).
- We would also continue to emit *doing it wrong* notices even when the responsible plugin is doing nothing wrong, in the sense that, in general, I don't think it's necessarily bad for other plugins to set options before `init` ... or, if it is, it probably isn't our job to police that—especially since those operations may be entirely unrelated to WooCommerce. It's only a problem at all because of the way the Features Controller is currently structured.

**Comment by barryhughes** [comment]
🤔 ...If we knew the option related to a feature, then I guess that alternative approach would work, because we could ignore other cases and only warn about early feature enablement/disablement. We can't (currently) do that, though, without first building the list of feature definitions, which is the nub of the problem. Solving that would require a larger change.

**Comment by coreymckrill** [comment]
Ah right. Ok, it seems like we should stick with this approach then.

**Pull request merged**